File: cfd_report/supabase_client.py
from __future__ import annotations

from datetime import date, timedelta
from typing import Any

import requests

from .config import Config


def _headers(cfg: Config) -> dict[str, str]:
    return {
        "apikey": cfg.supabase_anon_key,
        "Authorization": f"Bearer {cfg.supabase_anon_key}",
        "Content-Type": "application/json",
        "Accept": "application/json",
    }


def _upsert_headers(cfg: Config) -> dict[str, str]:
    h = _headers(cfg)
    h["Prefer"] = "resolution=merge-duplicates,return=minimal"
    return h


def check_cache(cfg: Config, target_date: date) -> dict[str, bool]:
    """检查 Supabase 中是否已有当日数据。"""
    today = target_date.isoformat()
    week = (target_date - timedelta(days=target_date.weekday())).isoformat()
    qkey = f"{target_date.year}-Q{(target_date.month - 1) // 3 + 1}"

    results: dict[str, bool] = {}
    for table, param in [
        ("worklog_daily", f"work_date=eq.{today}"),
        ("worklog_weekly", f"week_start=eq.{week}"),
        ("worklog_quarterly", f"quarter_key=eq.{qkey}"),
    ]:
        try:
            url = f"{cfg.supabase_url}/rest/v1/{table}?{param}&limit=1"
            r = requests.get(url, headers=_headers(cfg), timeout=10)
            r.raise_for_status()
            results[table.replace("worklog_", "")] = len(r.json()) > 0
        except Exception:
            results[table.replace("worklog_", "")] = False
    return results


def check_entries_cache(cfg: Config, work_date: str) -> bool:
    """检查 worklog_entries 是否已有指定日期数据。"""
    try:
        url = f"{cfg.supabase_url}/rest/v1/worklog_entries?work_date=eq.{work_date}&limit=1"
        response = requests.get(url, headers=_headers(cfg), timeout=10)
        response.raise_for_status()
        return len(response.json()) > 0
    except Exception:
        return False


def upsert_rows(cfg: Config, table: str, rows: list[dict[str, Any]]) -> int:
    if not rows:
        return 0
    conflict_keys = {
        "worklog_daily": "account_id,work_date",
        "worklog_weekly": "account_id,week_start",
        "worklog_quarterly": "account_id,quarter_key",
    }
    url = f"{cfg.supabase_url}/rest/v1/{table}"
    params = {"on_conflict": conflict_keys[table]} if table in conflict_keys else None
    resp = requests.post(
        url,
        params=params,
        json=rows,
        headers=_upsert_headers(cfg),
        timeout=20,
    )
    resp.raise_for_status()
    return len(rows)


def upsert_entry_rows(cfg: Config, rows: list[dict[str, Any]]) -> int:
    """upsert worklog_entries 行数据。"""
    if not rows:
        return 0
    url = f"{cfg.supabase_url}/rest/v1/worklog_entries"
    response = requests.post(
        url,
        params={"on_conflict": "account_id,work_date,issue_key"},
        json=rows,
        headers=_upsert_headers(cfg),
        timeout=20,
    )
    response.raise_for_status()
    return len(rows)


def fetch_daily(cfg: Config, work_date: str) -> list[dict[str, Any]]:
    url = f"{cfg.supabase_url}/rest/v1/worklog_daily?work_date=eq.{work_date}&order=display_name"
    r = requests.get(url, headers=_headers(cfg), timeout=10)
    r.raise_for_status()
    return r.json()


def fetch_weekly(cfg: Config, week_start: str) -> list[dict[str, Any]]:
    url = f"{cfg.supabase_url}/rest/v1/worklog_weekly?week_start=eq.{week_start}&order=display_name"
    r = requests.get(url, headers=_headers(cfg), timeout=10)
    r.raise_for_status()
    return r.json()


def fetch_quarterly(cfg: Config, quarter_key: str) -> list[dict[str, Any]]:
    url = f"{cfg.supabase_url}/rest/v1/worklog_quarterly?quarter_key=eq.{quarter_key}&order=display_name"
    r = requests.get(url, headers=_headers(cfg), timeout=10)
    r.raise_for_status()
    return r.json()


def _parse_total_count(content_range: str) -> int:
    if "/" not in content_range:
        return 0
    tail = content_range.rsplit("/", 1)[-1].strip()
    if tail == "*":
        return 0
    try:
        return int(tail)
    except ValueError:
        return 0


def count_high_estimate_entries(cfg: Config, work_date: str, threshold_sec: int) -> int:
    """统计指定日期内预估时长大于阈值的任务条数。"""
    url = (
        f"{cfg.supabase_url}/rest/v1/worklog_entries"
        f"?select=account_id"
        f"&work_date=eq.{work_date}"
        f"&estimated_sec=gt.{threshold_sec}"
        f"&limit=1"
    )
    headers = _headers(cfg)
    headers["Prefer"] = "count=exact"
    response = requests.get(url, headers=headers, timeout=10)
    response.raise_for_status()
    return _parse_total_count(response.headers.get("Content-Range", ""))


def fetch_high_estimate_entries(
    cfg: Config,
    work_date: str,
    threshold_sec: int,
    page_size: int,
    page_no: int,
) -> list[dict[str, Any]]:
    """分页查询指定日期内预估时长大于阈值的任务。"""
    safe_page_size = max(1, int(page_size))
    safe_page_no = max(1, int(page_no))
    offset = (safe_page_no - 1) * safe_page_size

    url = (
        f"{cfg.supabase_url}/rest/v1/worklog_entries"
        f"?select=account_id,display_name,work_date,issue_key,issue_summary,time_sec,estimated_sec"
        f"&work_date=eq.{work_date}"
        f"&estimated_sec=gt.{threshold_sec}"
        f"&order=estimated_sec.desc,issue_key.asc,account_id.asc"
        f"&limit={safe_page_size}"
        f"&offset={offset}"
    )
    response = requests.get(url, headers=_headers(cfg), timeout=10)
    response.raise_for_status()
    return response.json()

File: backend/src/main/java/com/cfd/client/SupabaseClient.java
package com.cfd.client;

import com.cfd.config.AppProperties;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.*;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.util.UriComponentsBuilder;

import java.time.LocalDate;
import java.util.*;

@Slf4j
@Component
@RequiredArgsConstructor
public class SupabaseClient {

    private final AppProperties props;
    private final RestTemplate restTemplate;
    private final ObjectMapper objectMapper;

    private HttpHeaders baseHeaders() {
        String key = props.getSupabase().getAnonKey();
        HttpHeaders headers = new HttpHeaders();
        headers.set("apikey", key);
        headers.set("Authorization", "Bearer " + key);
        headers.setAccept(List.of(MediaType.APPLICATION_JSON));
        return headers;
    }

    private HttpHeaders upsertHeaders() {
        HttpHeaders headers = baseHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        headers.set("Prefer", "resolution=merge-duplicates,return=minimal");
        return headers;
    }

    private HttpHeaders deleteHeaders() {
        HttpHeaders headers = baseHeaders();
        headers.set("Prefer", "return=minimal");
        return headers;
    }

    private String baseUrl(String table) {
        return props.getSupabase().getUrl() + "/rest/v1/" + table;
    }

    // ── Fetch ────────────────────────────────────────────────────────────

    public List<Map<String, Object>> fetchDailyRange(String dateFrom, String dateTo) {
        String base = baseUrl("worklog_daily")
                + "?work_date=gte." + dateFrom
                + "&work_date=lte." + dateTo
                + "&order=work_date.asc,display_name.asc";
        return fetchPaginated(base, 15_000, 2000);
    }

    public List<Map<String, Object>> fetchDaily(String workDate) {
        String base = baseUrl("worklog_daily")
                + "?work_date=eq." + workDate
                + "&order=display_name.asc";
        return fetchPaginated(base, 10_000, 2000);
    }

    public List<Map<String, Object>> fetchWeekly(String weekStart) {
        String base = baseUrl("worklog_weekly")
                + "?week_start=eq." + weekStart
                + "&order=display_name.asc";
        return fetchPaginated(base, 10_000, 2000);
    }

    public List<Map<String, Object>> fetchQuarterly(String quarterKey) {
        String base = baseUrl("worklog_quarterly")
                + "?quarter_key=eq." + quarterKey
                + "&order=display_name.asc";
        return fetchPaginated(base, 10_000, 2000);
    }

    public List<Map<String, Object>> fetchEntriesRange(String dateFrom, String dateTo) {
        String base = baseUrl("worklog_entries")
                + "?work_date=gte." + dateFrom
                + "&work_date=lte." + dateTo
                + "&order=work_date.asc,issue_key.asc,account_id.asc";
        return fetchPaginated(base, 20_000, 2000);
    }

    public List<Map<String, Object>> fetchEntriesRangeForMemberAggregate(String dateFrom, String dateTo) {
        String base = baseUrl("worklog_entries")
                + "?select=account_id,work_date,issue_key,estimated_sec"
                + "&work_date=gte." + dateFrom
                + "&work_date=lte." + dateTo;
        return fetchPaginated(base, 15_000, 2000);
    }

    private List<Map<String, Object>> fetchPaginated(String baseUrl, int timeoutMs, int pageSize) {
        List<Map<String, Object>> rows = new ArrayList<>();
        int offset = 0;

        while (true) {
            String url = baseUrl + "&limit=" + pageSize + "&offset=" + offset;
            HttpEntity<Void> entity = new HttpEntity<>(baseHeaders());
            ResponseEntity<List> resp = restTemplate.exchange(url, HttpMethod.GET, entity, List.class);

            List<Map<String, Object>> batch = Optional.ofNullable(resp.getBody()).orElse(List.of());
            rows.addAll(batch);
            if (batch.size() < pageSize) break;
            offset += pageSize;
        }

        return rows;
    }

    // ── Upsert ───────────────────────────────────────────────────────────

    public int upsertRows(String table, List<Map<String, Object>> rows) {
        if (rows.isEmpty()) return 0;

        Map<String, String> conflictKeys = Map.of(
                "worklog_daily", "account_id,work_date",
                "worklog_weekly", "account_id,week_start",
                "worklog_quarterly", "account_id,quarter_key"
        );

        String url = conflictKeys.containsKey(table)
                ? baseUrl(table) + "?on_conflict=" + conflictKeys.get(table)
                : baseUrl(table);

        HttpEntity<List<Map<String, Object>>> entity = new HttpEntity<>(rows, upsertHeaders());
        restTemplate.exchange(url, HttpMethod.POST, entity, Void.class);
        return rows.size();
    }

    public int upsertEntryRows(List<Map<String, Object>> rows) {
        if (rows.isEmpty()) return 0;

        String url = baseUrl("worklog_entries") + "?on_conflict=account_id,work_date,issue_key";
        HttpEntity<List<Map<String, Object>>> entity = new HttpEntity<>(rows, upsertHeaders());

        try {
            restTemplate.exchange(url, HttpMethod.POST, entity, Void.class);
            return rows.size();
        } catch (Exception e) {
            String msg = e.getMessage() != null ? e.getMessage().toLowerCase() : "";
            boolean isExtendedFieldError = msg.contains("project_key")
                    || msg.contains("project_name")
                    || msg.contains("status_name");

            if (!isExtendedFieldError) throw e;

            // Fallback: strip extended fields
            List<String> legacyCols = List.of("account_id", "display_name", "work_date",
                    "issue_key", "issue_summary", "time_sec", "estimated_sec", "updated_at");
            List<Map<String, Object>> legacyRows = rows.stream().map(row -> {
                Map<String, Object> r = new LinkedHashMap<>();
                for (String col : legacyCols) {
                    if (row.containsKey(col)) r.put(col, row.get(col));
                }
                return r;
            }).toList();

            HttpEntity<List<Map<String, Object>>> fallbackEntity = new HttpEntity<>(legacyRows, upsertHeaders());
            restTemplate.exchange(url, HttpMethod.POST, fallbackEntity, Void.class);
            return legacyRows.size();
        }
    }

    // ── Delete ───────────────────────────────────────────────────────────

    public void deleteByDateRange(String table, String dateFrom, String dateTo, String dateCol) {
        String url = baseUrl(table) + "?" + dateCol + "=gte." + dateFrom + "&" + dateCol + "=lte." + dateTo;
        HttpEntity<Void> entity = new HttpEntity<>(deleteHeaders());
        restTemplate.exchange(url, HttpMethod.DELETE, entity, Void.class);
    }

    public void deleteByWeekStarts(String table, List<String> weekStarts) {
        if (weekStarts.isEmpty()) return;
        String inClause = String.join(",", weekStarts);
        String url = baseUrl(table) + "?week_start=in.(" + inClause + ")";
        HttpEntity<Void> entity = new HttpEntity<>(deleteHeaders());
        restTemplate.exchange(url, HttpMethod.DELETE, entity, Void.class);
    }

    public void deleteByQuarterKey(String table, String quarterKey) {
        String url = baseUrl(table) + "?quarter_key=eq." + quarterKey;
        HttpEntity<Void> entity = new HttpEntity<>(deleteHeaders());
        restTemplate.exchange(url, HttpMethod.DELETE, entity, Void.class);
    }

    // ── Replace window ───────────────────────────────────────────────────

    public Map<String, Integer> replaceSyncWindowRows(
            String dateFrom, String dateTo, String quarterKey,
            List<Map<String, Object>> dailyRows,
            List<Map<String, Object>> weeklyRows,
            List<Map<String, Object>> quarterlyRows,
            List<Map<String, Object>> entryRows) {

        List<String> weekStarts = weekStartsInRange(dateFrom, dateTo);

        deleteByDateRange("worklog_entries", dateFrom, dateTo, "work_date");
        deleteByDateRange("worklog_daily", dateFrom, dateTo, "work_date");
        if (!weekStarts.isEmpty()) deleteByWeekStarts("worklog_weekly", weekStarts);
        deleteByQuarterKey("worklog_quarterly", quarterKey);

        return Map.of(
                "entries", upsertEntryRows(entryRows),
                "daily", upsertRows("worklog_daily", dailyRows),
                "weekly", upsertRows("worklog_weekly", weeklyRows),
                "quarterly", upsertRows("worklog_quarterly", quarterlyRows)
        );
    }

    public Map<String, Integer> replaceRangeWindowRows(
            String dateFrom, String dateTo,
            List<Map<String, Object>> dailyRows,
            List<Map<String, Object>> entryRows) {

        deleteByDateRange("worklog_entries", dateFrom, dateTo, "work_date");
        deleteByDateRange("worklog_daily", dateFrom, dateTo, "work_date");

        return Map.of(
                "entries", upsertEntryRows(entryRows),
                "daily", upsertRows("worklog_daily", dailyRows)
        );
    }

    // ── Helpers ──────────────────────────────────────────────────────────

    private List<String> weekStartsInRange(String dateFrom, String dateTo) {
        LocalDate start = LocalDate.parse(dateFrom);
        LocalDate end = LocalDate.parse(dateTo);
        if (start.isAfter(end)) return List.of();

        List<String> weeks = new ArrayList<>();
        LocalDate current = start.minusDays(start.getDayOfWeek().getValue() - 1);
        while (!current.isAfter(end)) {
            weeks.add(current.toString());
            current = current.plusWeeks(1);
        }
        return weeks;
    }
}
